Exception when loading proxy does not mark proxy as initialized

lstrojny opened the pull request:
Similar issue as https://github.com/doctrine/orm/issues/3147 but with a random exception, e.g. when the connection is closed.

greg0ire commented:
Please improve your commit message according to [the contributing guide](https://www.doctrine-project.org/contribute/index.html#working-on-topic-branches).

lstrojny committed:
Don’t mark proxy as initialized on exception

When an unspecified exception is triggered while initializing the proxy,
the proxy is marked as initialized even though the initialization didn’t
complete. The change catches any Throwable and resets the initialization
state if the exception is thrown.

lstrojny commented:
@greg0ire done

SenseException commented on tests/Doctrine/Tests/ORM/Proxy/ProxyFactoryTest.php:
```diff
@@ -139,6 +140,29 @@ public function testFailedProxyLoadingDoesNotMarkTheProxyAsInitialized(): void
+            ->method('load')
+            ->willThrowException(new Exception());
+
+        try {
+            $proxy->getDescription();
+            $this->fail('An exception was expected to be raised');
+        } catch (Exception $exception) {
+        }
```
```suggestion
        $this->expectException(Exception::class);
        $proxy->getDescription();
```

lstrojny replied:
Would make the test less precise: instead of expecting exactly `getDescription` to throw, the exception could happen anywhere, e.g. when calling `$proxy->__isInitialized()`

SenseException replied:
`expectException()` does exactly that. The method is more precise than the outdated annotation from the older PHPUnit versions.

SenseException commented on tests/Doctrine/Tests/ORM/Proxy/ProxyFactoryTest.php:
```diff
@@ -139,6 +140,29 @@ public function testFailedProxyLoadingDoesNotMarkTheProxyAsInitialized(): void
 
+    public function testExceptionWhenLoadingProxyDoesNotMarkTheProxyAsInitialized(): void
+    {
+        $persister = $this->getMockBuilder(BasicEntityPersister::class)->setMethods(['load'])->disableOriginalConstructor()->getMock();
```
```suggestion
        $persister = $this->getMockBuilder(BasicEntityPersister::class)
            ->setMethods(['load'])
            ->disableOriginalConstructor()
            ->getMock();
```

lstrojny replied:
Formatting is identical to the one above. I guess if the Doctrine projects wants a different way of formatting fluent interfaces, introducing a CS rule and applying it across the whole project would make more sense. What do you think?

greg0ire replied:
> I guess if the Doctrine projects wants a different way of formatting fluent interfaces

This is not specifically about fluent interfaces: this line is 135 characters long, while PSR recommends 120.

> introducing a CS rule and applying it across the whole project would make more sense. What do you think?

It would make sense, but although it sounds simple, it sure isn't, see http://journal.stuffwithstuff.com/2015/09/08/the-hardest-program-ive-ever-written/

lstrojny commented:
Bump. Could you have another look?

lstrojny commented:
Bump again

SenseException commented:
Please take a look at the open change request that needs to be referenced.
